### ISSUE TITLE
How to use with next.js (not custom express server) ?

### ISSUE THREAD

**Issue by trompx** [felixmosh/bull-board#124]
Hello,

All examples I've found use bull-board with a custom express server.
Is there a way to use it with next.js own server (cannot use app.use) for instance in a pages/bull-dashboard.js file ?

**Comment by SahanAmadoruge** [comment]
i'm looking for an answer for the same issue.

**Comment by noclat** [comment]
Here's how I made it work: 

```js
import Queue from 'bull';
import { setQueues, BullAdapter, router } from 'bull-board';
import nextConnect from 'next-connect';

setQueues([
  new BullAdapter(new Queue('main')),
]);

// tell Express to prepend all paths
router.use('/api/queues', router._router);

// tell Next.js to execute Express routes
// directly calling handler.use('/api/queues', router._router.handle); throws an error, idk why
const handler = nextConnect();
handler.use('/api/queues', (req, res, next) => {
  return router._router.handle(req, res, next);
});

export default handler;
```

A bit ugly but I've found a workaround. Underneath it's basically just an Express application. There's a way to execute an express route without launching the server: `app._router.handle` ([source](https://soyuka.me/call-framework-routes-programmatically/).

But it needs a bit more understanding when combining it with Next.js: we have to tell Next.js to route every subpath to that specific handler, and we have to tell the Express app to change base path for all its routes.

So to handle Next, we just need to create `/pages/api/queues/[[..path]].js`, where `[[..path]]` is the catch-all convention of Next.js. And keep in mind that as defined in the [Next.js doc](https://nextjs.org/docs/api-routes/introduction), API routes are only possible inside the `/pages/api` folder.

Then, we have to tell Express that all URLs will be prepended with `/api/queues`, the same way we'd do with Express, so using: `app.use('/path', handler)`. But the confusion here is that the whole app is exposed as `router` by the `bull-board` package. So when importing `import { router } from 'bull-board';`, `router` is actually a full Express app definition.

To make sure I don't mess with Express, I've also used [next-connect](https://www.npmjs.com/package/next-connect) to mimic the `next` argument.

**Disclaimer:** I'm still in the process of testing this. Haven't actually run a job and observed if it shows in the UI yet, but the UI is showing in full. This progress was worth sharing in itself.

**Comment by jafework** [comment]
@noclat Thanks! this worked, only change is `[[..path]].js` should be `[[...path]].js`

**Comment by dstotz** [comment]
Hey in case anyone else ends up here looking for how to use the newer `@bull-board` library with Next JS here is an updated snippet

```typescript
import { Queue } from 'bullmq';
import { createBullBoard } from '@bull-board/api';
import { BullMQAdapter } from '@bull-board/api/bullMQAdapter';
import { ExpressAdapter } from '@bull-board/express';
import nextConnect from 'next-connect';

const myQueue = new Queue('My Queue');

const serverAdapter = new ExpressAdapter();

createBullBoard({
  queues: [new BullMQAdapter(myQueue)],
  serverAdapter: serverAdapter,
});

const handler = nextConnect();

serverAdapter.setBasePath('/api/admin');
handler.use('/api/admin', serverAdapter.getRouter());

export default handler;
```

**Comment by cymen** [comment]
The above example but for regular Bull (not MQ):

```
import { createBullBoard } from '@bull-board/api';
import { BullAdapter } from '@bull-board/api/bullAdapter';
import { ExpressAdapter } from '@bull-board/express';
import nextConnect from 'next-connect';
import someQueue from "../../../../queues/some/queue";

const serverAdapter = new ExpressAdapter();

createBullBoard({
  queues: [new BullAdapter(someQueue)],
  serverAdapter: serverAdapter,
});

const handler = nextConnect();

serverAdapter.setBasePath('/api/admin/queues');
handler.use('/api/admin/queues', serverAdapter.getRouter());

export default handler;
```

I realized the path in the example directly above is `/api/admin` but the prior examples use `/api/admin/queues` so I used that folder structure as already had it setup from the prior examples (there is a confusing error if you get that part wrong).

**Comment by rdadoune** [comment]
The most current example above didn't work for me with `next-connect@^0.13.0` because of the following lines of code:
https://github.com/hoangvvo/next-connect/blob/cda00e1af285391119b4f9c2d60de7e6b9aa12db/src/index.js#L89-L95
I had to add an `all` so that not handlers are middleware:
```typescript
import { createBullBoard } from '@bull-board/api';
import { BullAdapter } from '@bull-board/api/bullAdapter';
import { ExpressAdapter } from '@bull-board/express';
import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
import nc from 'next-connect';
import someQueue from "../../../../queues/some/queue";

const serverAdapter = new ExpressAdapter();

createBullBoard({
  queues: [
    new BullAdapter(someQueue),
  ],
  serverAdapter: serverAdapter,
});

serverAdapter.setBasePath('/api/admin/queues');

const handler = nc<NextApiRequest, NextApiResponse>({
  attachParams: true,
});

handler
  .use('/api/admin/queues', serverAdapter.getRouter() as NextApiHandler)
  .all(() => ({}));

export default handler;


```

**Comment by caseypugh** [comment]
This is super helpful, thanks all!

Is anyone using Vercel for deployment? I'm using @rdadoune's solution which works great but when I deploy to Vercel and test the API endpoint it throws this error: `Cannot find module '@bull-board/ui/package.json'`. Full log: 

```
[GET] /api/admin/queues
16:15:28:24

2022-08-22T20:15:28.430Z	af893d15-e517-4c31-94e1-bdfb4bf12e11	
ERROR	Error: Cannot find module '@bull-board/ui/package.json'

Require stack:
- /var/task/node_modules/@bull-board/api/dist/src/index.js
- /var/task/.next/server/pages/api/admin/[[...path]].js
- /var/task/node_modules/next/dist/server/next-server.js
- /var/task/___next_launcher.cjs    
- at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)    
- at Function.resolve (node:internal/modules/cjs/helpers:108:19)    
- at eval (eval at createBullBoard (/var/task/node_modules/@bull-board/api/dist/src/index.js:13:47), <anonymous>:1:9)    
- at createBullBoard (/var/task/node_modules/@bull-board/api/dist/src/index.js:13:47)    
- at /var/task/.next/server/pages/api/admin/[[...path]].js:68:65 {  code: 'MODULE_NOT_FOUND',  
- requireStack: [    '/var/task/node_modules/@bull-board/api/dist/src/index.js',    '/var/task/.next/server/pages/api/admin/[[...path]].js',    '/var/task/node_modules/next/dist/server/next-server.js',    '/var/task/___next_launcher.cjs'  ]}RequestId: af893d15-e517-4c31-94e1-bdfb4bf12e11 Error: Runtime exited with error: exit status 1Runtime.ExitError
```

Everything works locally even when i do `next build && next start`. Just not on Vercel.

Using `@bull-board/express@^4.2.2` and `next-connect@^0.13.0`

**Comment by felixmosh** [comment]
The error comes from the source code which tried to load the package.json file, make sure it is deployed to Vercel as well

**Comment by caseypugh** [comment]
@felixmosh It seems to be more complex of an issue. I suspect [this line of code](https://github.com/felixmosh/bull-board/blob/master/packages/api/src/index.ts#L16) is not playing nice with Vercel. I _think_ it needs to specify the relative path for everything. I'll try and fork this and patch it ...

**Comment by felixmosh** [comment]
Please, open a separate issue for vercel

A try will help me a lot, thanx

**Comment by ravi-manoj** [comment]
> ```ts
> .all(() => ({}));
> ```



> The most current example above didn't work for me with `next-connect@^0.13.0` because of the following lines of code: https://github.com/hoangvvo/next-connect/blob/cda00e1af285391119b4f9c2d60de7e6b9aa12db/src/index.js#L89-L95 I had to add an `all` so that not handlers are middleware:
> 
> ```ts
> import { createBullBoard } from '@bull-board/api';
> import { BullAdapter } from '@bull-board/api/bullAdapter';
> import { ExpressAdapter } from '@bull-board/express';
> import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
> import nc from 'next-connect';
> import someQueue from "../../../../queues/some/queue";
> 
> const serverAdapter = new ExpressAdapter();
> 
> createBullBoard({
>   queues: [
>     new BullAdapter(someQueue),
>   ],
>   serverAdapter: serverAdapter,
> });
> 
> serverAdapter.setBasePath('/api/admin/queues');
> 
> const handler = nc<NextApiRequest, NextApiResponse>({
>   attachParams: true,
> });
> 
> handler
>   .use('/api/admin/queues', serverAdapter.getRouter() as NextApiHandler)
>   .all(() => ({}));
> 
> export default handler;
> ```

I'm new to next-js , I'm getting 404 error when I'm trying to reach "http://localhost:3000/api/admin"

**Comment by felixmosh** [comment]
Try the defined path, /api/admin/queues

**Comment by ravi-manoj** [comment]
![image](https://user-images.githubusercontent.com/125236527/225226280-26de543b-04c0-4ed0-8a20-041a7d841af0.png)
![image](https://user-images.githubusercontent.com/125236527/225226594-c2e12f02-6ff9-49a1-a23c-94412d27ffd0.png)
still getting 404 error

**Comment by felixmosh** [comment]
Your code shows that on the side of next the path should be /api/bull
Then on the queue side you should define the path

**Comment by ravi-manoj** [comment]
![image](https://user-images.githubusercontent.com/125236527/225237460-ae9aa8af-4ad3-4da4-a840-2dac3d2b4052.png)
![image](https://user-images.githubusercontent.com/125236527/225237609-9aa55db7-63af-4402-a903-8802bd82b872.png)

**Comment by ravi-manoj** [comment]
[cid:c3c3040c-21d6-4c61-bf0e-c44951558800]
[cid:7f6e31ed-c0be-4dcf-8384-b8aa5732a8ea]
________________________________

**Comment by felixmosh** [comment]
Please read this
https://github.com/felixmosh/bull-board/issues/526#issuecomment-1415442670

**Comment by theDanielJLewis** [comment]
Anyone have a solution for the Next.js 13 app router?

**Comment by samuelcole** [comment]
I got fed up trying to make this work with next-connect 1.0, so I just switched to express classic:

``` typescript
import { createBullBoard } from "@bull-board/api";
import { BullMQAdapter } from "@bull-board/api/bullMQAdapter";
import { ExpressAdapter } from "@bull-board/express";
import express from "express";
import queue from "@/queues/queue";

const basePath = "/api/bull-board";

const serverAdapter = new ExpressAdapter();
serverAdapter
  .setBasePath(basePath)
  .setViewsPath("../../node_modules/@bull-board/ui/dist");

createBullBoard({
  queues: [
      new BullMQAdapter(queue),
    ],
  serverAdapter: serverAdapter,
  options: { uiBasePath: "../../node_modules/@bull-board/ui", uiConfig: {} },
});

const handler = express().use(basePath, serverAdapter.getRouter());

export default handler;
```

**Comment by tikoflano** [comment]
I'm using NextJS 14 with the app router. I couldn't make it work with it, so I ended up using the old page router in parallel (you don't need anything extra for this).

My **pages/api/admin/[[...queues]].ts** file looks like this, and I can get the dashboard at **/api/admin/queues**

```js
import { Queue } from "bullmq";
import { createBullBoard } from "@bull-board/api";
import { BullAdapter } from "@bull-board/api/bullAdapter";
import { ExpressAdapter } from "@bull-board/express";
import express from "express";

const basePath = "/api/admin/queues";
const myQueue = new Queue("My Queue");

const serverAdapter = new ExpressAdapter();

createBullBoard({
  queues: [new BullAdapter(myQueue)],
  serverAdapter: serverAdapter,
});

serverAdapter.setBasePath(basePath);

const handler = express().use(basePath, serverAdapter.getRouter());

export default handler;
```

**Comment by fccoelho7** [comment]
> Anyone have a solution for the Next.js 13 app router?

You can use it with Hono

**Comment by mbrookson** [comment]
Thanks everyone for the examples in this thread. Thought I'd share my solution too based on this. I didn't need `next-connect`, just used raw `express`.

Also tried for a while but couldn't get this working with the Next.js app router. I think I got close but seemed like Next.js was throwing an error because express was trying to manipulate the request or response objects.

Anyway, here's a working solution _with some simple authorisation via `next-auth` and environment variables too_. Basically I validate only specific logged in users of my app can access bull-board.

```ts
import { env } from "@/env";
import { calendarEventsQueue } from "@/server/bullmq/queues";
import { createBullBoard } from "@bull-board/api";
import { BullMQAdapter } from "@bull-board/api/bullMQAdapter";
import { ExpressAdapter } from "@bull-board/express";
import express from "express";
import { type NextApiRequest, type NextApiResponse } from "next";
import { getSession } from "next-auth/react";

const app = express();
const basePath = "/api/bull";
const serverAdapter = new ExpressAdapter();

serverAdapter.setBasePath(basePath);

createBullBoard({
  queues: [
    new BullMQAdapter(calendarEventsQueue, {
      readOnlyMode: false,
      allowRetries: true,
    }),
  ],
  serverAdapter: serverAdapter,
});

const router = app.use(basePath, serverAdapter.getRouter());

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  const session = await getSession({ req });

  if (!session) {
    res.status(401).end();
    return;
  }

  if (!env.BULLMQ_ADMIN_USER_IDS?.includes(session.user.id)) {
    res.status(403).end();
    return;
  }

  return router(req, res);
}

```

**Comment by chihabhajji** [comment]
can this be used to track the progress of a queue that is running on a seperate backend in my nextjs app ?

**Comment by mbrookson** [comment]
@chihabhajji Yes that should work fine. It just requires a redis connection and you have to configure which queues it should look at.

**Comment by smoothdvd** [comment]
@fccoelho7 Thanks a lot!

There is my app router code

```
// src/app/api/queues/[[...index]]/route.ts

import { createBullBoard } from "@bull-board/api";
import { BullMQAdapter } from "@bull-board/api/bullMQAdapter";
import { HonoAdapter } from "@bull-board/hono";
import { getQueue } from "@/bullmq/queues";
import { Hono } from 'hono';
import { handle } from 'hono/vercel';
import { serveStatic } from '@hono/node-server/serve-static';

const app = new Hono();

// Create the Express adapter
const serverAdapter = new HonoAdapter(serveStatic);

// Create Bull Board with your queues
createBullBoard({
  queues: [new BullMQAdapter(getQueue())],
  serverAdapter,
});

// Configure the server adapter
serverAdapter.setBasePath("/api/queues");
app.route('/api/queues', serverAdapter.registerPlugin());

export const GET = handle(app);
export const POST = handle(app);
```

Furthermore, I can create a middleware to protect the Bull Board.

**Comment by theDanielJLewis** [comment]
THANK YOU, @smoothdvd and @fccoelho7!

**Comment by theDanielJLewis** [comment]
Ugh. Unfortunately, it seems hono 4.6.9+ or @hono/node-server 1.13.5+ break this.

**Comment by theDanielJLewis** [comment]
@smoothdvd, are you able to retry or clear anything from your bull-board? It keeps trying to send post requests to `/api/queues/api/queues…`, but displaying data from the queues is working fine.

**Comment by smoothdvd** [comment]
> @smoothdvd, are you able to retry or clear anything from your bull-board? It keeps trying to send post requests to `/api/queues/api/queues…`, but displaying data from the queues is working fine.

I remember that I can retry the failed jobs and clean up the jobs in version 6.3.

**Comment by theDanielJLewis** [comment]
Ah ha!

The only thing I needed to add to @smoothdvd's code was this additional line:

```js
export const PUT = handle(app);
```

And now it properly handles retries and removals.

So the full code looks like this:

```js
// src/app/api/queues/[[...index]]/route.ts

import { createBullBoard } from "@bull-board/api";
import { BullMQAdapter } from "@bull-board/api/bullMQAdapter";
import { HonoAdapter } from "@bull-board/hono";
import { getQueue } from "@/bullmq/queues";
import { Hono } from 'hono';
import { handle } from 'hono/vercel';
import { serveStatic } from '@hono/node-server/serve-static';

const app = new Hono();

// Create the Express adapter
const serverAdapter = new HonoAdapter(serveStatic);

// Create Bull Board with your queues
createBullBoard({
  queues: [new BullMQAdapter(getQueue())],
  serverAdapter,
});

// Configure the server adapter
serverAdapter.setBasePath("/api/queues");
app.route('/api/queues', serverAdapter.registerPlugin());

export const GET = handle(app);
export const POST = handle(app);
export const PUT = handle(app);
```

I guess Bull Board changed to using PUT requests at some point.

Now to figure out how to lock this down to be accessible only to admin accounts.

**Comment by theDanielJLewis** [comment]
I figured this out to use my `adminOnly()` function. Replace the last three lines with the following (and import missing modules).

```js
const handleWithAuth = async (request: Request) => {
  await adminOnly();
  return handle(app)(request);
};
export const GET = handleWithAuth;
export const POST = handleWithAuth;
export const PUT = handleWithAuth;
```

**Comment by smoothdvd** [comment]
> I figured this out to use my `adminOnly()` function. Replace the last three lines with the following (and import missing modules).
> 
> ```js
> const handleWithAuth = async (request: Request) => {
>   await adminOnly();
>   return handle(app)(request);
> };
> export const GET = handleWithAuth;
> export const POST = handleWithAuth;
> export const PUT = handleWithAuth;
> ```

Sorry, I forgot to add PUT handle in the above code

**Comment by theDanielJLewis** [comment]
Do you know how many hours … 🤣

I actually temporarily gave up and went back to a clunky Express approach and proxying it.

**Comment by smoothdvd** [comment]
@theDanielJLewis Have you encountered this error in the Next.js production environment?

```
api-1          |  ⨯ Error: Cannot find module '@bull-board/ui/package.json'
api-1          | Require stack:
api-1          | - /app/.next/server/app/bull-board/[[...index]]/route.js
api-1          | - /app/node_modules/.pnpm/next@15.0.3_@babel+core@7.26.0_@opentelemetry+api@1.9.0_react-dom@19.0.0-rc-66855b96-20241106_k3dcxjmnq54knewkfd4y77cxki/node_modules/next/dist/server/require.js
api-1          | - /app/node_modules/.pnpm/next@15.0.3_@babel+core@7.26.0_@opentelemetry+api@1.9.0_react-dom@19.0.0-rc-66855b96-20241106_k3dcxjmnq54knewkfd4y77cxki/node_modules/next/dist/server/next-server.js
api-1          | - /app/node_modules/.pnpm/next@15.0.3_@babel+core@7.26.0_@opentelemetry+api@1.9.0_react-dom@19.0.0-rc-66855b96-20241106_k3dcxjmnq54knewkfd4y77cxki/node_modules/next/dist/server/next.js
api-1          | - /app/server.js
api-1          |     at <unknown> (/app/node_modules/.pnpm/next@15.0.3_@babel+core@7.26.0_@opentelemetry+api@1.9.0_react-dom@19.0.0-rc-66855b96-20241106_k3dcxjmnq54knewkfd4y77cxki/node_modules/next/dist/server/require-hook.js:55:36)
api-1          |     at eval (.next/server/app/bull-board/[[...index]]/route.js:1:37708)
api-1          |     at createBullBoard (/app/.next/server/app/bull-board/[[...index]]/route.js:1:37708)
api-1          |     at 76541 (/app/.next/server/app/bull-board/[[...index]]/route.js:1:29750)
api-1          |     at t (/app/.next/server/webpack-runtime.js:1:143)
api-1          |     at __webpack_exec__ (/app/.next/server/app/bull-board/[[...index]]/route.js:6528:246463)
api-1          |     at <unknown> (/app/.next/server/app/bull-board/[[...index]]/route.js:6528:246578)
api-1          |     at t.X (/app/.next/server/webpack-runtime.js:1:1514) {
api-1          |   code: 'MODULE_NOT_FOUND',
api-1          |   requireStack: [Array]
api-1          | }
```

**Comment by theDanielJLewis** [comment]
@smoothdvd, I think I've encountered an error like that. And the solution is simple, albeit maybe annoying: ditch pnpm. The way it handles module folders apparently causes trouble with Next.js 15. So try npm, yarn, or bun.
